idaes/apps/grid_integration/tracker.py

```diff
@@ -0,0 +1,1747 @@
+
+        def prod_cost_fun(m,j,h):
+            return m.min_load_cost[j] * m.on_off[j,h] \
+            + sum(m.F[j,l]*m.power_segment[j,h,l] for l in m.SEGMENTS)
+        m.prod_cost_approx = pyo.Expression(m.UNITS,m.HOUR,rule = prod_cost_fun)
+
+        # start up costs
+        def start_cost_fun(m,j,h):
+            return m.start_up_cost[j]*m.start_up[j,h]
+        m.start_up_cost_expr = pyo.Expression(m.UNITS,m.HOUR,rule = start_cost_fun)
+
+        # total cost
+        def tot_cost_fun(m,j,h):
+            return m.prod_cost_approx[j,h] + m.start_up_cost_expr[j,h]
+        m.tot_cost = pyo.Expression(m.UNITS,m.HOUR,rule = tot_cost_fun)
+
+        ## Objective
+        def exp_revenue_fun(m):
+            return sum(m.P_total[j,h]*m.DAM_price[h]- m.tot_cost[j,h] for h in m.HOUR for j in m.UNITS)
+        m.exp_revenue = pyo.Objective(rule = exp_revenue_fun,sense = pyo.maximize)
+        m.exp_revenue.deactivate()
+
+        return m
+
+    @staticmethod
+    def _add_tracking_constraints_and_objectives(m, ramping_weight = 20):
+
+        # add params for objective weights
+        m.cost_weight = pyo.Param(initialize = 1, mutable = False)
+        m.ramping_weight = pyo.Param(initialize = ramping_weight, mutable = False)
+
+        ## USE L1 norm
+        # add a slack variable for ramping
+        m.slack_var_ramp = pyo.Var(m.UNITS,m.HOUR,within = pyo.NonNegativeReals)
+
+        # add constraints for ramping
+        def abs_ramp_con_fun1(m,j,h):
+            if h == 0:
+                return -m.slack_var_ramp[j,h]<=m.P_T[j,h] - m.pre_P_T[j]
+            else:
+                return -m.slack_var_ramp[j,h]<=m.P_T[j,h] - m.P_T[j,h-1]
+        m.abs_ramp_con1 = pyo.Constraint(m.UNITS,m.HOUR,rule = abs_ramp_con_fun1)
+
+        def abs_ramp_con_fun2(m,j,h):
+            if h == 0:
+                return m.slack_var_ramp[j,h]>=m.P_T[j,h] - m.pre_P_T[j]
+            else:
+                return m.slack_var_ramp[j,h]>=m.P_T[j,h] - m.P_T[j,h-1]
+        m.abs_ramp_con2 = pyo.Constraint(m.UNITS,m.HOUR,rule = abs_ramp_con_fun2)
+
+        def power_abs_least(m):
+            return m.ramping_weight * sum(m.slack_var_ramp[j,h] for h in m.HOUR for j in m.UNITS)\
+                   + m.cost_weight * sum(m.tot_cost[j,h] for h in m.HOUR for j in m.UNITS)
+        m.abs_dev = pyo.Objective(rule = power_abs_least,sense = pyo.minimize)
+
+        return
+
+    def build_tracking_model(self,plan_horizon,ramping_weight = 20,segment_number=4,model_mode = '1a'):
+
+        m = self._build_model(self.model_data,plan_horizon,segment_number)
+
+        self._add_tracking_constraints_and_objectives(m, ramping_weight = ramping_weight)
+        self._add_slack_vars(m)
+        self._add_slack_constraints(m)
+        self._add_min_slack_objective(m)
+        self._add_slack_bounding_constraints(m)
+
+        self.switch_model_mode(m,mode = model_mode)
+
+        return m
+
+    @staticmethod
+    def switch_model_mode(m,mode = '1a'):
+
```

Comment:
This appears to be duplicate with the bidder. Happy to talk about ways to reduce redundant code.